x/auth/middleware/branch_store.go

```diff
@@ -8,47 +8,77 @@ import (
 	tmtypes "github.com/tendermint/tendermint/types"
 )
 
-type branchStoreHandler struct {
-	next tx.Handler
+// WithAnteBranch creates a store branch (cache-wrapped multistore) for Antehandlers.
+//
+// Usage:
+// beginBranch, endBranch := WithAnteBranch()
+//
+// ComposeMiddlewares(
+//   beginBranch,
+//   // some middlewares representing antehandlers
+//   endBranch, // will write to state before all middlewares below
+//   // some other middlewares
+// )
+func WithAnteBranch() (tx.Middleware, tx.Middleware) {
+	// We'll store the original sdk.Context multistore, as well as a
+	// cache-wrapped multistore, inside the sdk.Context's Values (it's just
+	// a simple key-value map). We define those keys here, and pass them to the
+	// middlewares.
+	originalMSKey := sdk.ContextKey("ante-original-ms")
```

Comment:
nit: might be helpful for code traceability and testability if these context keys were defined as package/file level constants